src/main/java/me/zeroeightsix/kami/util/graphics/font/KamiFontRenderer.kt

```diff
@@ -0,0 +1,222 @@
+                posX = 0.0
+            } else {
+                val pos1 = Vec2d(posX, posY)
+                val pos2 = pos1.add(charInfo.width.toDouble(), charInfo.height.toDouble())
+                val texPos1 = Vec2d(charInfo.posX.toDouble(), charInfo.posY.toDouble()).divide(TEXTURE_WIDTH.toDouble(), chunk.textureHeight.toDouble())
+                val texPos2 = texPos1.add(Vec2d(charInfo.width.toDouble(), charInfo.height.toDouble()).divide(TEXTURE_WIDTH.toDouble(), chunk.textureHeight.toDouble()))
+
+                drawQuad(pos1, pos2, texPos1, texPos2)
+                posX += charInfo.width - 1f
+            }
+        }
+
+        glPopMatrix()
+        if (lighting) glEnable(GL_LIGHTING)
+        glEnable(GL_ALPHA_TEST)
+        GlStateUtils.cull(true)
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
+        glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_LOD_BIAS, 0.0f)
+        glColor4f(1f, 1f, 1f, 1f)
+    }
+
+    private fun getShadowColor(color: ColorHolder) = ColorHolder((color.r * 0.2f).toInt(), (color.g * 0.2f).toInt(), (color.b * 0.2f).toInt(), (color.a * 0.9f).toInt())
+
+    private fun drawQuad(pos1: Vec2d, pos2: Vec2d, texPos1: Vec2d, texPos2: Vec2d) {
+        buffer.begin(GL_TRIANGLE_STRIP, DefaultVertexFormats.POSITION_TEX)
+        buffer.pos(pos1.x, pos1.y, 0.0).tex(texPos1.x, texPos1.y).endVertex()
+        buffer.pos(pos2.x, pos1.y, 0.0).tex(texPos2.x, texPos1.y).endVertex()
+        buffer.pos(pos1.x, pos2.y, 0.0).tex(texPos1.x, texPos2.y).endVertex()
+        buffer.pos(pos2.x, pos2.y, 0.0).tex(texPos2.x, texPos2.y).endVertex()
+        tessellator.draw()
+    }
+
+    @JvmOverloads
+    fun getFontHeight(scale: Float = 1f): Float {
+        return glyphArray[0].fontHeight * 0.25f * scale
+    }
+
+    @JvmOverloads
+    fun getStringWidth(text: String, scale: Float = 1f): Float {
+        var width = 0f
+        currentVariant = glyphArray[0]
+        for ((index, char) in text.withIndex()) {
+            if (checkStyleCode(text, index)) continue
+            width += currentVariant.getCharInfo(char).width.minus(1f)
+        }
+
+        return width * 0.28f * scale
+    }
+
+    private fun checkStyleCode(text: String, index: Int): Boolean {
+        if (text.getOrNull(index - 1) == '§') return true
+
+        if (text.getOrNull(index) == '§') {
+            when (text.getOrNull(index + 1)) {
+                TextProperties.Style.REGULAR.codeChar -> currentVariant = glyphArray[0]
+                TextProperties.Style.BOLD.codeChar -> currentVariant = glyphArray[1]
+                TextProperties.Style.ITALIC.codeChar -> currentVariant = glyphArray[2]
+            }
+            currentColor = when (text.getOrNull(index + 1)) {
```

Comment:
this should be cached somewhere using a delegated lazy

Reply:
No, because you need to update it as the font renderer encounters color code